app/models/agents/s3_agent.rb

```diff
@@ -0,0 +1,206 @@
+module Agents
+  class S3Agent < Agent
+    include FormConfigurable
+    include FileHandling
+
+    emits_file_pointer!
+    no_bulk_receive!
+
+    default_schedule 'every_1h'
+
+    gem_dependency_check { defined?(Aws::S3) }
+
+    description do
+      <<-MD
+        The S3Agent can watch a bucket for changes or emit an event for every file in that bucket. When receiving events it writes the data into a file on S3.
+
+        #{'## Include `aws-sdk-core` in your Gemfile to use this Agent!' if dependencies_missing?}
+
+        `mode` must be present and either `read` or `write`, in `read` mode the agent checks the S3 bucket for changed files, with `write` it writes received events to a file in the bucket.
+
+        ### Universal options
+
+        To use credials for the `access_key` and `access_key_secret` use the liquid `credential` tag like so `{% credential name-of-credential %}`
+
+        Select the `region` in which the bucket was created.
+
+        ### Reading
+
+        Then `watch` is set to `true` the S3Agent will watch the specified `bucket` for changes. An event will be emitted for every detected change.
+
+        Then `watch` is set to `false` the agent will emit an event for every file in the directory on each sheduled run.
+
+        #{emitting_file_handling_agent_description}
+
+        ### Writing
+
+        Specify the filename to use in `filename`, Liquid interpolation is possible to change the name per event.
+
+        Use [JSONPath](http://goessner.net/articles/JsonPath/) to select the which part of the received event should be written in `data_path`.
```

Comment:
Should be possible, but I am not sure how much slower it would be with large payloads. The transformation could be done using a `EventFormattingAgent`.